src/backend/base/langflow/api/v1/mcp.py

```diff
@@ -0,0 +1,295 @@
+import asyncio
+import json
+import logging
+import traceback
+from contextvars import ContextVar
+from typing import Annotated
+from uuid import UUID, uuid4
+
+import pydantic
+from anyio import BrokenResourceError
+from fastapi import APIRouter, Depends, Request
+from fastapi.responses import StreamingResponse
+from mcp import types
+from mcp.server import NotificationOptions, Server
+from mcp.server.sse import SseServerTransport
+from sqlmodel import select
+from starlette.background import BackgroundTasks
+
+from langflow.api.v1.chat import build_flow
+from langflow.api.v1.schemas import InputValueRequest
+from langflow.graph import Graph
+from langflow.services.auth.utils import get_current_active_user
+from langflow.services.database.models import Flow, User
+from langflow.services.deps import get_db_service, get_session
+
+logger = logging.getLogger(__name__)
+if False:
+    logger.setLevel(logging.DEBUG)
+    if not logger.handlers:
+        handler = logging.StreamHandler()
+        handler.setLevel(logging.DEBUG)
+        formatter = logging.Formatter("[%(asctime)s][%(levelname)s] %(message)s")
+        handler.setFormatter(formatter)
+        logger.addHandler(handler)
+
+    # Enable debug logging for MCP package
+    mcp_logger = logging.getLogger("mcp")
+    mcp_logger.setLevel(logging.DEBUG)
+    if not mcp_logger.handlers:
+        mcp_logger.addHandler(handler)
+
+    logger.debug("MCP module loaded - debug logging enabled")
+
+router = APIRouter(prefix="/mcp", tags=["mcp"])
+
+server = Server("langflow-mcp-server")
+
+# Create a context variable to store the current user
+current_user_ctx: ContextVar[User] = ContextVar("current_user_ctx")
+
+
+def json_schema_from_flow(flow: Flow) -> dict:
```

Comment:
Take a look at `test_graph_state_model_json_schema` for a way of getting the Graph's json schema.

Reply:
I had a look and that test uses:


    graph_state_model: BaseModel = create_state_model_from_graph(graph)()
    json_schema = graph_state_model.model_json_schema(mode="serialization")
    
    
but that blows up with:

```
[2024-12-19 09:49:04,002][ERROR] Task exception was never retrieved
future: <Task finished name='Task-1991' coro=<<async_generator_athrow without __name__>()> exception=IllegalStateChangeError("Method 'close()' can't be called here; method '_connection_for_bind()' is already in progress and this would cause an unexpected state change to <SessionTransactionState.CLOSED: 5>")>
Traceback (most recent call last):
  File "/home/tato/Desktop/langflow/src/backend/base/langflow/services/database/service.py", line 152, in with_async_session
    yield session
  File "/home/tato/Desktop/langflow/src/backend/base/langflow/services/deps.py", line 153, in get_session
    yield session
GeneratorExit
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/home/tato/.local/share/uv/python/cpython-3.12.6-linux-x86_64-gnu/lib/python3.12/asyncio/tasks.py", line 316, in __step_run_and_handle_result
    result = coro.throw(exc)
             ^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/src/backend/base/langflow/services/deps.py", line 152, in get_session
    async with get_db_service().with_async_session() as session:
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/.local/share/uv/python/cpython-3.12.6-linux-x86_64-gnu/lib/python3.12/contextlib.py", line 231, in __aexit__
    await self.gen.athrow(value)
  File "/home/tato/Desktop/langflow/src/backend/base/langflow/services/database/service.py", line 151, in with_async_session
    async with AsyncSession(self.async_engine, expire_on_commit=False) as session:
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/sqlalchemy/ext/asyncio/session.py", line 1080, in __aexit__
    await asyncio.shield(task)
  File "/home/tato/.local/share/uv/python/cpython-3.12.6-linux-x86_64-gnu/lib/python3.12/asyncio/futures.py", line 291, in __await__
    yield self  # This tells Task to wait for completion.
    ^^^^^^^^^^
  File "/home/tato/.local/share/uv/python/cpython-3.12.6-linux-x86_64-gnu/lib/python3.12/asyncio/tasks.py", line 385, in __wakeup
    future.result()
  File "/home/tato/.local/share/uv/python/cpython-3.12.6-linux-x86_64-gnu/lib/python3.12/asyncio/futures.py", line 203, in result
    raise self._exception.with_traceback(self._exception_tb)
  File "/home/tato/.local/share/uv/python/cpython-3.12.6-linux-x86_64-gnu/lib/python3.12/asyncio/tasks.py", line 314, in __step_run_and_handle_result
    result = coro.send(None)
             ^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/sqlalchemy/ext/asyncio/session.py", line 1025, in close
    await greenlet_spawn(self.sync_session.close)
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 190, in greenlet_spawn
    result = context.switch(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/sqlalchemy/orm/session.py", line 2518, in close
    self._close_impl(invalidate=False)
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/sqlalchemy/orm/session.py", line 2587, in _close_impl
    transaction.close(invalidate)
  File "<string>", line 2, in close
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/sqlalchemy/orm/state_changes.py", line 121, in _go
    raise sa_exc.IllegalStateChangeError(
sqlalchemy.exc.IllegalStateChangeError: Method 'close()' can't be called here; method '_connection_for_bind()' is already in progress and this would cause an unexpected state change to <SessionTransactionState.CLOSED: 5> (Background on this error at: https://sqlalche.me/e/20/isce)
[2024-12-19 09:49:09,689][ERROR] Error in listing tools: Cannot generate a JsonSchema for core_schema.CallableSchema
For further information visit https://errors.pydantic.dev/2.7/u/invalid-for-json-schema
[2024-12-19 09:49:09,691][ERROR] Traceback (most recent call last):
  File "/home/tato/Desktop/langflow/src/backend/base/langflow/api/v1/mcp.py", line 137, in handle_list_tools
    inputSchema=json_schema_from_flow(flow),
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/src/backend/base/langflow/api/v1/mcp.py", line 65, in json_schema_from_flow
    json_schema = graph_state_model.model_json_schema(mode="serialization")
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/main.py", line 433, in model_json_schema
    return model_json_schema(
           ^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 2241, in model_json_schema
    return schema_generator_instance.generate(cls.__pydantic_core_schema__, mode=mode)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 412, in generate
    json_schema: JsonSchemaValue = self.generate_inner(schema)
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 551, in generate_inner
    json_schema = current_handler(schema)
                  ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py", line 36, in __call__
    return self.handler(core_schema)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 508, in handler_func
    json_schema = generate_for_schema_type(schema_or_field)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 1807, in definitions_schema
    return self.generate_inner(schema['schema'])
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 551, in generate_inner
    json_schema = current_handler(schema)
                  ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py", line 36, in __call__
    return self.handler(core_schema)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 525, in new_handler_func
    json_schema = js_modify_function(schema_or_field, current_handler)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/main.py", line 652, in __get_pydantic_json_schema__
    return handler(core_schema)
           ^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py", line 36, in __call__
    return self.handler(core_schema)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 525, in new_handler_func
    json_schema = js_modify_function(schema_or_field, current_handler)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 220, in modify_model_json_schema
    json_schema = handler(schema_or_field)
                  ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py", line 36, in __call__
    return self.handler(core_schema)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 508, in handler_func
    json_schema = generate_for_schema_type(schema_or_field)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 1392, in model_schema
    json_schema = self.generate_inner(schema['schema'])
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 551, in generate_inner
    json_schema = current_handler(schema)
                  ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py", line 36, in __call__
    return self.handler(core_schema)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 508, in handler_func
    json_schema = generate_for_schema_type(schema_or_field)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 1484, in model_fields_schema
    json_schema = self._named_required_fields_schema(named_required_fields)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 1295, in _named_required_fields_schema
    field_json_schema = self.generate_inner(field).copy()
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 551, in generate_inner
    json_schema = current_handler(schema)
                  ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py", line 36, in __call__
    return self.handler(core_schema)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 543, in new_handler_func
    json_schema = js_modify_function(schema_or_field, current_handler)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 1719, in set_computed_field_metadata
    json_schema = handler(schema)
                  ^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py", line 36, in __call__
    return self.handler(core_schema)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 508, in handler_func
    json_schema = generate_for_schema_type(schema_or_field)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 1374, in computed_field_schema
    return self.generate_inner(schema['return_schema'])
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 551, in generate_inner
    json_schema = current_handler(schema)
                  ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py", line 36, in __call__
    return self.handler(core_schema)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 525, in new_handler_func
    json_schema = js_modify_function(schema_or_field, current_handler)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/main.py", line 652, in __get_pydantic_json_schema__
    return handler(core_schema)
           ^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py", line 36, in __call__
    return self.handler(core_schema)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 525, in new_handler_func
    json_schema = js_modify_function(schema_or_field, current_handler)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 220, in modify_model_json_schema
    json_schema = handler(schema_or_field)
                  ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py", line 36, in __call__
    return self.handler(core_schema)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 508, in handler_func
    json_schema = generate_for_schema_type(schema_or_field)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 1392, in model_schema
    json_schema = self.generate_inner(schema['schema'])
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 551, in generate_inner
    json_schema = current_handler(schema)
                  ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py", line 36, in __call__
    return self.handler(core_schema)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 508, in handler_func
    json_schema = generate_for_schema_type(schema_or_field)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 1484, in model_fields_schema
    json_schema = self._named_required_fields_schema(named_required_fields)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 1295, in _named_required_fields_schema
    field_json_schema = self.generate_inner(field).copy()
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 551, in generate_inner
    json_schema = current_handler(schema)
                  ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py", line 36, in __call__
    return self.handler(core_schema)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 543, in new_handler_func
    json_schema = js_modify_function(schema_or_field, current_handler)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 1719, in set_computed_field_metadata
    json_schema = handler(schema)
                  ^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py", line 36, in __call__
    return self.handler(core_schema)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 508, in handler_func
    json_schema = generate_for_schema_type(schema_or_field)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 1374, in computed_field_schema
    return self.generate_inner(schema['return_schema'])
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 551, in generate_inner
    json_schema = current_handler(schema)
                  ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py", line 36, in __call__
    return self.handler(core_schema)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 525, in new_handler_func
    json_schema = js_modify_function(schema_or_field, current_handler)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/main.py", line 652, in __get_pydantic_json_schema__
    return handler(core_schema)
           ^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py", line 36, in __call__
    return self.handler(core_schema)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 525, in new_handler_func
    json_schema = js_modify_function(schema_or_field, current_handler)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 220, in modify_model_json_schema
    json_schema = handler(schema_or_field)
                  ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py", line 36, in __call__
    return self.handler(core_schema)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 508, in handler_func
    json_schema = generate_for_schema_type(schema_or_field)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 1392, in model_schema
    json_schema = self.generate_inner(schema['schema'])
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 551, in generate_inner
    json_schema = current_handler(schema)
                  ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py", line 36, in __call__
    return self.handler(core_schema)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 508, in handler_func
    json_schema = generate_for_schema_type(schema_or_field)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 997, in function_before_schema
    return self._function_schema(schema)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 981, in _function_schema
    return self.generate_inner(schema['schema'])
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 551, in generate_inner
    json_schema = current_handler(schema)
                  ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py", line 36, in __call__
    return self.handler(core_schema)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 508, in handler_func
    json_schema = generate_for_schema_type(schema_or_field)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 1484, in model_fields_schema
    json_schema = self._named_required_fields_schema(named_required_fields)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 1295, in _named_required_fields_schema
    field_json_schema = self.generate_inner(field).copy()
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 551, in generate_inner
    json_schema = current_handler(schema)
                  ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py", line 36, in __call__
    return self.handler(core_schema)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 543, in new_handler_func
    json_schema = js_modify_function(schema_or_field, current_handler)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 2184, in json_schema_update_func
    json_schema = {**handler(core_schema_or_field), **json_schema_update}
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py", line 36, in __call__
    return self.handler(core_schema)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 508, in handler_func
    json_schema = generate_for_schema_type(schema_or_field)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 1363, in model_field_schema
    return self.generate_inner(schema['schema'])
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 551, in generate_inner
    json_schema = current_handler(schema)
                  ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py", line 36, in __call__
    return self.handler(core_schema)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 508, in handler_func
    json_schema = generate_for_schema_type(schema_or_field)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 1082, in nullable_schema
    inner_json_schema = self.generate_inner(schema['schema'])
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 551, in generate_inner
    json_schema = current_handler(schema)
                  ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py", line 36, in __call__
    return self.handler(core_schema)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 508, in handler_func
    json_schema = generate_for_schema_type(schema_or_field)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 834, in callable_schema
    return self.handle_invalid_for_json_schema(schema, 'core_schema.CallableSchema')
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tato/Desktop/langflow/.venv/lib/python3.12/site-packages/pydantic/json_schema.py", line 2152, in handle_invalid_for_json_schema
    raise PydanticInvalidForJsonSchema(f'Cannot generate a JsonSchema for {error_info}')
pydantic.errors.PydanticInvalidForJsonSchema: Cannot generate a JsonSchema for core_schema.CallableSchema
For further information visit https://errors.pydantic.dev/2.7/u/invalid-for-json-schema
```

because of callables so I'm not sure this approach will work, suggestions?

Reply:
That may be related to updates to pydantic. If it works like it is we can keep it like that but we should definitely create a default function for this.

Reply:
pulled it out to the flows helper